### ISSUE TITLE
Footplus

### ISSUE THREAD

**Issue by osenikalunga-droid** [jac99/FootAndBall#12]
import cv2
import numpy as np
import streamlit as st
import tempfile
from ultralytics import YOLO
from sklearn.cluster import KMeans

class AnalyseurFootballComplet:
    def __init__(self, chemin_modele='yolov8n.pt'):
        # Télécharge automatiquement le modèle pré-entraîné de base lors du premier lancement
        self.model = YOLO(chemin_modele)
        self.kmeans = None
        self.couleurs_equipes = {}
        self.noms_groupes = {}

    def extraire_couleur_maillot(self, image, box):
        x1, y1, x2, y2 = box
        hauteur = y2 - y1
        h_debut = int(y1 + hauteur * 0.15)
        h_fin = int(y1 + hauteur * 0.45)
        
        crop = image[h_debut:h_fin, x1:x2]
        if crop.size == 0:
            return None
            
        crop_hsv = cv2.cvtColor(crop, cv2.COLOR_BGR2HSV)
        return cv2.mean(crop_hsv)[:3]

    def calibrer_acteurs(self, chemin_video, nb_frames=120):
        cap = cv2.VideoCapture(chemin_video)
        caracteristiques = []
        compteur = 0

        while cap.isOpened() and compteur < nb_frames:
            ret, frame = cap.read()
            if not ret:
                break
            results = self.model(frame, classes=[0, 32], verbose=False)
            for box in results.boxes:
                if box.conf > 0.6:
                    coords = map(int, box.xyxy)
                    couleur = self.extraire_couleur_maillot(frame, list(coords))
                    if couleur is not None:
                        caracteristiques.append(couleur)
            compteur += 1
        cap.release()

        if len(caracteristiques) >= 3:
            self.kmeans = KMeans(n_clusters=3, random_state=42, n_init=10)
            self.kmeans.fit(caracteristiques)
            
            # Palette d'affichage BGR (Bleu pour l'équipe A, Jaune pour l'équipe B, Vert pour l'arbitre)
            self.couleurs_equipes = {
                0: (255, 0, 0),    
                1: (0, 255, 255),  
                2: (0, 255, 0)     
            }
            self.noms_groupes = {0: "Equipe A", 1: "Equipe B", 2: "Arbitre"}
            return True
        return False

    def calculer_distance(self, p1, p2):
        return np.sqrt((p1[0] - p2[0])**2 + (p1[1] - p2[1])**2)

    def analyser_match(self, chemin_video, conteneur_video, conteneur_stats):
        cap = cv2.VideoCapture(chemin_video)
        largeur = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
        hauteur = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
        total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))

        possession = {"Equipe A": 0, "Equipe B": 0, "Neutre": 0}
        derniere_equipe_possession = "Neutre"
        barre_progression = st.progress(0)
        compteur_frame = 0

        while cap.isOpened():
            ret, frame = cap.read()
            if not ret:
                break

            compteur_frame += 1
            results = self.model(frame, classes=[0, 32], verbose=False)
            
            coordonnees_ballon = None
            joueurs_detectes = []

            for box in results.boxes:
                x1, y1, x2, y2 = map(int, box.xyxy)
                conf = box.conf
                cls = int(box.cls)

                if conf > 0.45:
                    centre = (int((x1 + x2) / 2), int((y1 + y2) / 2))
                    
                    if cls == 32:  
                        coordonnees_ballon = centre
                        # Tracer le ballon en rouge
                        cv2.rectangle(frame, (x1, y1), (x2, y2), (0, 0, 255), 2)
                        cv2.putText(frame, "Ballon", (x1, y1 - 7), cv2.FONT_HERSHEY_SIMPLEX, 0.4, (0, 0, 255), 2)
                    elif cls == 0:  
                        couleur = self.extraire_couleur_maillot(frame, [x1, y1, x2, y2])
                        label = "Joueur"
                        couleur_box = (255, 255, 255)
                        
                        if couleur is not None and self.kmeans is not None:
                            groupe = self.kmeans.predict([couleur])[0]
                            label = self.noms_groupes[groupe]
                            couleur_box = self.couleurs_equipes[groupe]
                        
                        joueurs_detectes.append({"label": label, "pieds": (centre[0], y2), "box": (x1, y1, x2, y2), "couleur": couleur_box})

            # Algorithme de proximité du ballon pour la possession
            equipe_actuelle_possession = "Neutre"
            if coordonnees_ballon is not None and joueurs_detectes:
                distance_min = float('inf')
                for joueur in joueurs_detectes:
                    if joueur["label"] != "Arbitre":
                        dist = self.calculer_distance(joueur["pieds"], coordonnees_ballon)
                        if dist < distance_min:
                            distance_min = dist
                            if dist < 70:  # Seuil de proximité en pixels
                                equipe_actuelle_possession = joueur["label"]

                if equipe_actuelle_possession != "Neutre":
                    derniere_equipe_possession = equipe_actuelle_possession

            possession[derniere_equipe_possession] += 1

            # Rendu visuel des rectangles joueurs
            for joueur in joueurs_detectes:
                x1, y1, x2, y2 = joueur["box"]
                cv2.rectangle(frame, (x1, y1), (x2, y2), joueur["couleur"], 2)
                cv2.putText(frame, joueur["label"], (x1, y1 - 7), cv2.FONT_HERSHEY_SIMPLEX, 0.4, joueur["couleur"], 2)

            # Calcul des pourcentages de possession
            total_possession = sum(possession.values()) if sum(possession.values()) > 0 else 1
            pct_a = (possession["Equipe A"] / total_possession) * 100
            pct_b = (possession["Equipe B"] / total_possession) * 100
            
            # Incrustation du tableau des scores directement sur la vidéo
            cv2.rectangle(frame, (10, 10), (320, 80), (0, 0, 0), -1)
            cv2.putText(frame, f"Possession A: {pct_a:.1f}%", (20, 40), cv2.FONT_HERSHEY_SIMPLEX, 0.6, (255, 255, 255), 2)
            cv2.putText(frame, f"Possession B: {pct_b:.1f}%", (20, 65), cv2.FONT_HERSHEY_SIMPLEX, 0.6, (255, 255, 255), 2)

            # Envoi de la vidéo traitée vers l'interface web
            frame_rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
            conteneur_video.image(frame_rgb, channels="RGB", use_container_width=True)
            
            # Mise à jour du widget de statistiques
            conteneur_stats.markdown(f"""
            ### 📊 Analyse Tactique Live
            * **Possession {self.noms_groupes.get(0, 'A')} :** {pct_a:.1f}%
            * **Possession {self.noms_groupes.get(1, 'B')} :** {pct_b:.1f}%
            * **Acteurs détectés à l'écran :** {len(joueurs_detectes)}
            """)

            barre_progression.progress(compteur_frame / total_frames)

        cap.release()
        barre_progression.empty()

# --- BLOC INTERFACE UTILISATEUR (STREAMLIT) ---
st.set_page_config(page_title="IA Football Analytics", layout="wide")
st.title("⚽ Application Tactique : Détection Automatique des Acteurs")
st.write("Outil universitaire complet de détection automatique (Joueurs, Arbitres, Ballons) et de statistiques tactiques.")

col1, col2 = st.columns([3, 1])

with col1:
    conteneur_video = st.empty()
    st.info("💡 Chargez une vidéo dans la barre latérale gauche pour initialiser le système.")

with col2:
    conteneur_stats = st.empty()
    conteneur_stats.write("### 📊 Statistiques Live\nAucune vidéo active.")

st.sidebar.header("🔧 Panneau de Contrôle")
fichier_video = st.sidebar.file_uploader("Importer le fichier du match (.mp4)", type=["mp4", "mov", "avi"])

if fichier_video is not None:
    tfile = tempfile.NamedTemporaryFile(delete=False)
    tfile.write(fichier_video.read())
    chemin_video = tfile.name

    if st.sidebar.button("▶️ Démarrer l'Analyse"):
        with st.spinner("Phase 1/2 : Calibrage de l'IA sur la couleur des maillots..."):
            analyseur = AnalyseurFootballComplet()
            succes_calibrage = analyseur.calibrer_acteurs(chemin_video)
            
        if succes_calibrage:
            st.sidebar.success("✅ Modèle calibré automatiquement sur le match !")
            with st.spinner("Phase 2/2 : Inférence vidéo et suivi en temps réel..."):
                analyseur.analyser_match(chemin_video, conteneur_video, conteneur_stats)
            st.success("🎉 Analyse vidéo finalisée !")
        else:
            st.sidebar.error("❌ Échec du calibrage. Assurez-vous que des joueurs soient bien visibles sur les premières secondes de la vidéo.")
